Error codes checkup and rustdoc test fix

GuillaumeGomez opened the pull request:
This PR does a few things:

 * fix how rustdoc checks that an error code has been thrown (it only checked for "E0XXX" so if it appeared in the output because the file has it in its name or wherever, it passed the test, which was incorrect)
 * fix the failing code examples that weren't throwing the expected error code

rust-highfive commented:
The job `x86_64-gnu-llvm-7` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=17480&view=logs&jobId=0230d20d-e0df-5b62-a6ec-fad0f9c3c3d5) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/17480/logs/65), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/17480/logs/65)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
2020-01-04T00:45:09.2620133Z ##[command]git remote add origin https://github.com/rust-lang/rust
2020-01-04T00:45:10.0570101Z ##[command]git config gc.auto 0
2020-01-04T00:45:10.0577246Z ##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
2020-01-04T00:45:10.0582873Z ##[command]git config --get-all http.proxy
2020-01-04T00:45:10.0586443Z ##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/67850/merge:refs/remotes/pull/67850/merge
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Mark-Simulacrum commented:
I will review once https://github.com/rust-lang/rust/pull/67837 has been merged and this is rebased atop of it (also, looks like there's a CI failure).

GuillaumeGomez commented:
Forgot to run rustfmt on the new checker... Let's wait for #67837 now. :)

GuillaumeGomez commented:
@Mark-Simulacrum #67837 has been merged. :)

Mark-Simulacrum commented on src/bootstrap/test.rs:
```diff
@@ -704,6 +704,48 @@ impl Step for RustdocUi {
+            .env("CFG_RELEASE_CHANNEL", &builder.config.channel)
+            .env("RUSTDOC_REAL", builder.rustdoc(self.compiler))
+            .env("RUSTDOC_CRATE_VERSION", builder.rust_version())
+            .env("RUSTC_BOOTSTRAP", "1");
```
Do you actually need most of these? I imagine that you could directly pass `builder.rustdoc(compiler)` as the argument to the tool and omit ~all of these variables.

Mark-Simulacrum commented on src/tools/error-codes/main.rs:
```diff
@@ -0,0 +1,57 @@
+use std::process::{exit, Command};
+
+fn get_error_codes<P: AsRef<Path>>(folder_path: P) -> Vec<String> {
+    let mut ret = Vec::with_capacity(10);
```
Please drop the capacity here or add a comment as to why 10. AFAICT, it's not true that we expect exactly 10 elements, and regardless we shouldn't really be optimizing this code.

Mark-Simulacrum commented on src/tools/error-codes/main.rs:
```diff
@@ -0,0 +1,57 @@
+        if !path.is_file() {
+            continue;
+        }
+        ret.push(format!("{}", path.display()));
```
You never seem to use the String-ness of the error codes, so maybe you can just return `Vec<PathBuf>`? If we do use it let's call display() there and not here.

GuillaumeGomez replied:
I use it as an argument and print it afterwards. Both cases require strings.

ollie27 commented on src/bootstrap/test.rs:
```diff
@@ -704,6 +704,39 @@ impl Step for RustdocUi {
 }
 
+#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
+pub struct ErrorCodesCheck {
```
There is already a step to run rustdoc tests on the error index: https://github.com/rust-lang/rust/blob/175631311716d7dfeceec40d2587cde7142ffa8c/src/bootstrap/test.rs#L1393

Does it not work properly?

GuillaumeGomez replied:
This is not for the same purpose. `ErrorIndex` generates it, but it doesn't test that the code examples inside the explanations are compiling (or not) as expected.

ollie27 replied:
Yes it does. It uses `error_index_generator` to concatenate all of the error code docs into a single `error-index.md` file and then runs `rustdoc --test` on that: https://github.com/rust-lang/rust/blob/175631311716d7dfeceec40d2587cde7142ffa8c/src/bootstrap/test.rs#L1411-L1436

GuillaumeGomez replied:
And I can tell that it doesn't work since a lot of error codes I fixed in this PR weren't detected.

Mark-Simulacrum replied:
But *why* doesn't it work? If we have existing code that overlaps this one, ideally we'd fix the pre-existing code, rather than replace it wholesale (or at least, we should make sure that we're replacing it, rather than just also adding atop it).

GuillaumeGomez replied:
This is a good question. Let me check again.

ollie27 commented on src/bootstrap/test.rs:
```diff
@@ -704,6 +704,39 @@ impl Step for RustdocUi {
+    fn run(self, builder: &Builder<'_>) {
+        builder.ensure(compile::Std { compiler: self.compiler, target: self.compiler.host });
+
+        let mut cmd = builder.tool_cmd(Tool::ErrorCodesCheck);
```
Does this really need to be done using a new tool? Rustbuild can already run `rustdoc --test` on all `.md` files in a directory using the `DocTest` step.

Either way this will be slower than the current `ErrorIndex` step because it will have to call `rustdoc --test` once for each error code rather than just once for all of them.

GuillaumeGomez replied:
You need to write a tool in any case to import the markdown files. So instead of having everything at once, at least we get a very precise error output. Also, the run is a bit longer, but nothing much bigger.

GuillaumeGomez commented:
Ping @Mark-Simulacrum

Mark-Simulacrum commented:
I think @ollie27's comments around error-index-generator need to get resolved and then I can take another look (or not, if it turns out we're already doing everything necessary here).

GuillaumeGomez commented:
@ollie27 was absolutely right: adding this new check was useless. I discovered this issue, believing that the change of the error explanations format somehow messed with the run of `rustdoc --test` so I fixed rustdoc and adding the new (and useless) checker. Luckily, it was detected and I removed it. Thanks a lot @ollie27!

I think this is ready now.

Mark-Simulacrum commented:
Please update the PR description (some of it seems a bit outdated). After that, r=me.

GuillaumeGomez commented:
@bors: r=Mark-Simulacrum

bors commented:
:pushpin: Commit 5b7d64e9473216dc8650ca5f910100389ab91726 has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 5b7d64e9473216dc8650ca5f910100389ab91726 -->
<!-- homu: {"type":"Approved","sha":"5b7d64e9473216dc8650ca5f910100389ab91726","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 5b7d64e9473216dc8650ca5f910100389ab91726 with merge bf84eb538fd16743240434b3e837b36c35719fee...
<!-- homu: {"type":"BuildStarted","head_sha":"5b7d64e9473216dc8650ca5f910100389ab91726","merge_sha":"bf84eb538fd16743240434b3e837b36c35719fee"} -->

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=18320)
Approved by: Mark-Simulacrum
Pushing bf84eb538fd16743240434b3e837b36c35719fee to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=18320"},"merge_sha":"bf84eb538fd16743240434b3e837b36c35719fee"} -->
